Add progress status for partition rebalances (#140)
strimzi/proposals: 088-rebalance-progress-status.md

# Partition Rebalance Progress Status

This proposal introduces a new feature to monitor the progression of an ongoing partition rebalance executed by a Strimzi-managed Cruise Control instance via a `KafkaRebalance` custom resource. 

## Current Situation

At this time, Strimzi users are able to execute partition rebalances via `KafkaRebalance` custom resources but can only monitor the progression of those partition rebalances in two ways:

- Manually querying the Cruise Control REST API endpoint directly. 
- Inspecting the logs of the Cruise Control instance. 

Unfortunately, neither of these methods are particularly user-friendly.

## Motivation

Information surrounding the progress of an executing partition rebalance is useful for planning future cluster operations. 
Knowing things like how much time an ongoing partition rebalance has left to take and how much data an ongoing partition rebalance has left to transfer helps users understand the cost of an ongoing partition rebalance.
This information helps users decide whether they should continue or cancel an ongoing rebalance, and know when future operations will be able to be safely executed.

> Further, having this information readily available and easily accessible via Kubernetes primitives, allows users and third-party tools like the Kubernetes CLI or Strimzi Console to easily track the progression of a partition rebalance.

Reviewer: "Strimzi Console"? I think you mean the StreamsHub Console.

## Proposal

> This proposal extends the status section of the `KafkaRebalance` custom resource to include a `progress` section with a nested `rebalanceProgressConfigMap` field that references a `ConfigMap` that contains information related to an ongoing partition rebalance.

Reviewer 2: Do we need a new CM? Can't we use the existing one from the proposal?

Author: We were considering using the existing ` ConfigMap`, "afterBeforeLoadConfigMap" to store this progress information, but were concerned the additional data would contribute to hitting the 1 MB `ConfigMap` limit sooner. It is not so much of an issue for the constant "non-verbose" executor state information we plan on providing as part of this proposal. However, if we were to extend the feature to provide the variable "verbose" executor state information in the future, it would increase the chance of hitting the limit for larger production clusters that have a larger number of brokers and partitions.

If we have no requests/plans for providing "verbose" executor state information in the future, I don't see much of a problem of storing the information in the existing `ConfigMap`. At the least, it would simplify the proposal implementation . Any thoughts @ppatierno @tomncooper?

Reviewer 3: As mentioned before, I am not sure we'll never have the support for "verbose" so maybe we could think about the present and not the future. From this perspective, using the same ConfigMap seems to be reasonable.
Anyway you can keep the `rebalanceProgressConfigMap` field pointing to that ConfigMap.
When/if one day we have support for "verbose", that field will just point to a different ConfigMap. It should not be a big issue for users.

Reviewer 2: You don't have to use the same CM just because I asked about it.

Reviewer 3: No I just think it's a good compromise. Anyway let's see what @kyguy @tomncooper think?

Author: Discussed offline with Paolo and Tom,  since the progress information is constant, we can safely add it to the existing `ConfigMap` maintained for and tied to the `KafkadRebalance` resource. This keeps `KafkaRebalance` information organized in one place, simplifies the proposal implementation, and has insignificant impact on the storage of the `ConfigMap`. Refactored and added this note to the proposal.

apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: Rebalancing | Stopped [1]
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress: [1]
    rebalanceProgressConfigMap: my-rebalance-progress [2]
```
[1] The `progress` section will be visible during the KafkaRebalance `Rebalancing` and  `Stopped` states.

[2] The `ConfigMap` containing information related to the ongoing partition rebalance, generated with the name "<kafka_rebalance_resource_name>-progress".

In the `ConfigMap`, we will include the following fields:

> - **estimatedTimeToCompletion**: The estimated amount time it will take in minutes until partition rebalance is complete.

Reviewer 4: Like in `optimizationResult`, I think that it would be good to add a unit suffix to this key, i.e. `estimatedCompletionTimeInMinutes`. IMO, we should follow this pattern in general, and avoid using unit suffixes in values. This would also make the following formulas self explanatory.

> - **percentageDataMovementComplete**: The percentage of the data movement of the partition rebalance that is completed e.g. values in the range [0-100]%

Reviewer 4: Should this be named `completedDataMovementPercentage`, similar to `monitoredPartitionsPercentage` in `optimizationResult`?

- **executorState**: The “non-verbose” JSON payload from the `/kafkacruisecontrol/state?substates=executor` endpoint.

The progress `ConfigMap` of an inter-broker partition rebalance would look like the following:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-rebalance-progress
  …
data:
  estimatedTimeToCompletion: 5m [1]
  percentageDataMovementComplete: 80% [2]
  executorState: [3]
  {
   "abortingPartitions":0,
   "averageConcurrentPartitionMovementsPerBroker":5,
   "finishedDataMovement":0,
   "maximumConcurrentPartitionMovementsPerBroker":5,
   "minimumConcurrentPartitionMovementsPerBroker":5,
   "numFinishedPartitionMovements":0,
   "numInProgressPartitionMovements":0,
   "numPendingPartitionMovements":20,
   "numTotalPartitionMovements":20,
   "state":"INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS",
   "totalDataToMove":0,
   "triggeredSelfHealingTaskId":"",
   "triggeredTaskReason":"No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
   "triggeredUserTaskId":"0230d401-6a36-430e-9858-fac8f2edde93"
  }
```

> [1] The estimated time it will take the rebalance to complete based on the average rate of data transfer.

Reviewer 4: Which time unit? There are other places in which the time unit is not defined.

[2] The percentage complete of the ongoing rebalance in the range [0-100]%

[3] The “non-verbose” JSON payload from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint.

### Field values per `KafkaRebalance` State

For the initial implementation, we will focus on providing the progress information in the following `KafkaRebalance` states:

- `Rebalancing`
- `Stopped`

These are the states where this progress information will be able to be most accurately calculated and most useful for users. 

> We could provide this information for other states as well, such as the `ProposalReady` and `Ready` states, but it is not completely necessary, nor is it trivial. 

Reviewer: I still argue that having a lower bound for the estimated time of a `KafkaRebalance` in the `ProposalReady` state, based on the proposed data to move and the throttle settings would be a useful thing to have. It would obviously not be accurate as there is not disk throughput available. 

But it is still a useful guide and helps users gauge the impact of a proposed rebalance, which data-to-move values alone don't give.

Further discussion on the inclusion of the progress information for these other states can be found in the [Future Improvements](#future-improvements) section near the bottom of this proposal.

> All the information required for estimating the values of `estimatedTimeToCompletion` and `percentageDataMovementComplete` fields can be derived from either the Cruise Control server configurations or the [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) REST API endpoint.

Author: Updated "'or" -> "and" now that the `ProposalReady` state is being implemented as part of the proposal. This statement now makes more sense since the `ProposalReady` state estimations depend on the information from the Cruise Control server configurations, while the other states depend on the information from the `/kafkacruisecontrol/state?substates=executor`.

However, the actual formula used to produce values for these fields depends on the state of the `KafkaRebalance` resource.
Checkout the example in the [Field: executorState](#field-executorstate) section to see where the fields used in the formulas below come from.

### Field: `estimatedTimeToCompletion`

The estimated time it will take a rebalance to complete based on the average rate of data transfer.

The formulas used to calculate field value per `KafkaRebalance` state:
 
#### State: `Rebalancing`

$$
\text{rate} = \frac{\text{finishedDataMovement}\_{[1]}}{\text{taskTriggerTime}\_{[2]} - \text{currentTime}}
$$

$$
\text{estimatedTimeToCompletion} = \frac{\text{totalDataToMove}\_{[3]} - \text{finishedDataMovement}\_{[1]}}{\text{rate}}
$$

Notes

> - [1] `finishedDataMovement` is the number of megabytes already moved by rebalance, provided by [/kafkacruisecontrol/state?substates=executor](#field-executorstate) REST API endpoint.

Reviewer: You don't need the `[1]` in the formulas. You state the full name of the variable anyway so they don't help.

In the event that Cruise Control runs into an error when rebalancing, the operator will transition the `KafkaRebalance` resource to the `NotReady` state, remove the `progress` section, and delete the progress `ConfigMap`.
In the event that the Cruise Control REST API returns an error or fails to respond to the operator when querying the `/kafkacruisecontrol/state?substates=executor` endpoint during a rebalance, the operator will add a condition entry for the `Rebalancing` type with the message "Failed to retrieve rebalance progress" but leave the existing progress section and progress `ConfigMap` as is.

When Cruise Control state retrievel failed, the `KafkaRebalance` resource would be updated as follows:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  ...
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: Rebalancing

>     message: "Failed to retrieve rebalance progress"   

Reviewer 2: This should be probably a separate warning condition and not be part of the rebalancing condition as it is not clear what it means really?

Author: Yes, you are right,  just updated this to a "Warning condition"

  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress:
    rebalanceProgressConfigMap: my-rebalance-progress
```

> ### Future Improvements

Reviewer 2: Is this chapter empty? Or do you need to fix the headers from here down?

Author: Removed

> # Partition Rebalance Progress Status

Reviewer 3: Should be more "Cluster rebalance progress status" (or rebalancing) ... not sure if "partition" (even using the singular) sounds really fine because it's about a cluster rebalancing and about moving one or more partitions across the cluster.

> # Partition Rebalance Progress Status

Reviewer: ```suggestion
# Adding progress updates for Cruise Control rebalances
```

This proposal introduces a new feature to monitor the progression of an ongoing partition rebalance executed by a Strimzi-managed Cruise Control instance via a `KafkaRebalance` custom resource. 

## Current Situation

At this time, Strimzi users are able to execute partition rebalances via `KafkaRebalance` custom resources but can only monitor the progression of those partition rebalances in two ways:

> - Manually querying the Cruise Control REST API endpoint directly. 

Reviewer: Don't we lock these down with a Network Policy? Would the user have to alter the default set up to get access?

Author: Yes there are a couple of things that would need special configuration to enable a user to access the CC REST API directly

Reviewer 2: I thought that the user API is now supported after the proposal to manage the users?

Please keep in mind that the network policies are there to allow the operator access and not just to prevent users from accessing it. (In environments where the network plugin denies-by-default, we need to create the network policy to make sure the operator can talk with the CC API)

Author: > I thought that the user API is now supported after the proposal to manage the users?

It is. What what I mean by "special configuration" here is that in order to hit the Cruise Control API endpoint directly, a person will need to alter the default Cruise Control setup - Add Cruise Control API user to authenticate, create `NetworkPolicy` to access Cruise Control API port from outside pod in denies-by-default environments, etc

Author: It requires creating a custom REST API user or disabling HTTP Basic authentication, I'll add a note here.

A custom `NetworkPolicy` is only required in denies-by-default environments, mentioning the custom API user requirement should drive home the point without being too pedantic.

- Inspecting the logs of the Cruise Control instance. 

Unfortunately, neither of these methods are particularly user-friendly.

## Motivation

Information surrounding the progress of an executing partition rebalance is useful for planning future cluster operations. 
Knowing things like how much time an ongoing partition rebalance has left to take and how much data an ongoing partition rebalance has left to transfer helps users understand the cost of an ongoing partition rebalance.
This information helps users decide whether they should continue or cancel an ongoing rebalance, and know when future operations will be able to be safely executed.

> Further, having this information readily available and easily accessible via Kubernetes primitives, allows users and third-party tools like the Kubernetes CLI or StreamsHub Console to easily track the progression of a partition rebalance.

Reviewer 3: How many people in the Strimzi community would know about the "StreamsHub Console"? Maybe it deserves a link to the repo or not mentioning it at all?

## Proposal

This proposal extends the status section of the `KafkaRebalance` custom resource to include a `progress` section with a nested `rebalanceProgressConfigMap` field.

> This field will reference the `KafkaRebalance`'s existing `ConfigMap`, which will be enhanced to contain information related to an ongoing partition rebalance.

Reviewer 3: "the `KafkaRebalance`'s existing `ConfigMap`" ... maybe we should mention this ConfigMap beforehand to explain what it contains currently and from where it is already referenced (see `afterBeforeLoadConfigMap` field).

The `progress` section of the `KafkaRebalance` resource will look like the following:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: ProposalReady | Rebalancing | Stopped | Ready [1]
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress: [1]
    rebalanceProgressConfigMap: my-rebalance [2]
```

> [1] The `progress` section will be visible during the `Ready`, `Rebalancing`, `Stopped` and `Ready` states.

Reviewer 3: `PausedReconciliation` is not a valid rebalancing state.

> [1] The `progress` section will be visible during the `Ready`, `Rebalancing`, `Stopped` and `Ready` states.

Author: > Why Stopped but not PausedReconciliation or Not Ready? should we explain here?

The `PausedReconciliation` and  `NotReady` states are not related to the rebalance operation but more to the proposal genration. Therefore, these states don't have any rebalance progress associated with them

> [1] The `progress` section will be visible during the `Ready`, `Rebalancing`, `Stopped` and `Ready` states.

Reviewer 3: > The PausedReconciliation and NotReady states are not related to the rebalance operation but more to the proposal genration.

Well, actually even during a rebalancing you can get errors from CC and the `KafkaRebalance` ends in the `NotReady` state, right? But `PausedReconciliation` is not a valid rebalancing state at all.

> [1] The `progress` section will be visible during the `Ready`, `Rebalancing`, `Stopped` and `Ready` states.

Author: > Well, actually even during a rebalancing you can get errors from CC and the KafkaRebalance ends in the NotReady state, right?

Ah, yes you are right, any configuration or rebalance errors will put `KafkaRebalance` resource in `NotReady` state. Sorry @PaulRMellor, I was incorrect, `NotReady` should be supported as well for the same reason the `Stopped` state is supported, to show how far the rebalance got before it failed. That was a nice spot!

> But PausedReconciliation is not a valid rebalancing state at all.

Is that because it is related to the resource and not the rebalance itself? What determines whether it is a valid rebalancing state? I am confused because there is an enum for `PausedReconciliation` listed in the `KafkaRebalanceState` class [1]

[1] https://github.com/strimzi/strimzi-kafka-operator/blob/0.45.0/api/src/main/java/io/strimzi/api/kafka/model/rebalance/KafkaRebalanceState.java#L77

> [1] The `progress` section will be visible during the `Ready`, `Rebalancing`, `Stopped` and `Ready` states.

Reviewer 3: > Is that because it is related to the resource and not the rebalance itself? What determines whether it is a valid rebalancing state? I am confused because there is an enum for PausedReconciliation listed in the KafkaRebalanceState class [1]

But it's `ReconciliationPaused` not `PausedReconciliation`! :-P
Joking apart ... I was trying to defend myself because I totally missed this state in the rebalance FSM :-D

> [1] The `progress` section will be visible during the `Ready`, `Rebalancing`, `Stopped` and `Ready` states.

Author: I didn't know about it either until Paul mentioned it!

[2] The `ConfigMap` containing information related to the ongoing partition rebalance

In the `ConfigMap`, we will add the following fields:
- **estimatedTimeToCompletionInMinutes**: The estimated amount time it will take in minutes until partition rebalance is complete.
- **completedDataMovementPercentage**: The percentage of the data movement of the partition rebalance that is completed e.g. values in the range [0-100]%

> - **executorState**: The “non-verbose” JSON payload from the `/kafkacruisecontrol/state?substates=executor` endpoint.

Reviewer 3: I would not provide a detailed list of everything, maybe just a link to the OpenAPI definition of it on the Cruise Control repo?

> - **executorState**: The “non-verbose” JSON payload from the `/kafkacruisecontrol/state?substates=executor` endpoint.

Reviewer: Yeah, I agree you should give a brief summary of what this is and link to the OpenAPI def upstream.

> - **executorState**: The “non-verbose” JSON payload from the `/kafkacruisecontrol/state?substates=executor` endpoint.

Author: Added the link and text suggested by Paul which should be fair compromise

> - **executorState**: The “non-verbose” JSON payload from the `/kafkacruisecontrol/state?substates=executor` endpoint.

Author: Added

Since the progress information is constant, we can safely add it to the existing `ConfigMap` maintained for and tied to the `KafkadRebalance` resource.
This keeps `KafkaRebalance` information organized in one place, simplifies the proposal implementation, and has insignificant impact on the storage of the `ConfigMap`.

The enhanced `ConfigMap` of an inter-broker partition rebalance will look like the following:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-rebalance
  …
data:

>   estimatedTimeToCompletionInMinutes: 5m [1]

Reviewer 2: How is the estimation counted? Is it reliable? How much is it affected by the issues with unknown real network capacity?

Author: > How is the estimation counted? 

Depends on the `KafkaRebalance` state, the specific details per state are in the "Field: `estimatedTimeToCompletionInMinutes`" section of the proposal.

>  Is it reliable? 

In general, yes. The value for the `Stopped` and `Ready` states are hardcoded and the value for the `Rebalancing` state is based on the average rate of data transfer and easily calculated without the need of any user or capacity settings. The only state that could be potentially problematic is the estimation for the `ProposalReady` state which relies on accurate network capacity configuration from the user. 

> How much is it affected by the issues with unknown real network capacity?

If the default or user-configured network capacity is largely different from the real network capacity, the estimation for the `ProposalReady` state could be inaccurate. If the real network capacity is underestimated, the rebalance could take much less time than than `estimatedTimeToCompletionInMinutes` to complete. If the real network capacity is overestimated, the rebalance could take much more time than the `estimatedTimeToCompletionInMinutes` to complete. The former case wouldn't be an issue as we advertise  `estimatedTimeToCompletionInMinutes`  to be a theoretical minimum estimation in the `ProposalReady` state. However, the latter case would be an issue since the `estimatedTimeToCompletionInMinutes` value wouldn't be a theoretical minimum.

To avoid issues like this, the current plan is to document the users must provide accurate network capacity settings to have accurate `estimatedTimeToCompletionInMinutes` values in the `ProposalReady` state. We already documented that users must provide accurate network capacity settings to have accurate rebalances based on network capacity and distribution anyway.

Reviewer 2: > To avoid issues like this, the current plan is to document the users must provide accurate network capacity settings to have accurate estimatedTimeToCompletionInMinutes values in the ProposalReady state. We already documented that users must provide accurate network capacity settings to have accurate rebalances based on network capacity and distribution anyway.

I'm not sure this is a real solution. Do you really believe they configure the `accurate network capacity`? Do we even know how would they find out the `accurate network capacity`? Or will we solve it on paper but 99% or users will have it miscofigured and these numbers will be useless?

Author: >  Do you really believe they configure the accurate network capacity?

Users that are serious about their network resource usage and distribution do

>  Do we even know how would they find out the accurate network capacity?

I imagine they would use K8s CNI plugins or network performance benchmark tools

> Or will we solve it on paper but 99% or users will have it configured and these numbers will be useless?
> I'm not sure this is a real solution.

For users that have network capacity properly configured, this feature is still useful. I admit that I don't know how many Strimzi users configure their network capacity settings but I would like to believe that those that are doing it are doing it accurately. In addition to the network capacity documentation, what if we were to only include this estimation in the `ProposalReady` state for users that users explicitly configured their network capacity settings?

Reviewer 3: > In addition to the network capacity documentation, what if we were to only include this estimation in the ProposalReady state for users that explicitly configured their network capacity settings? Would that be a more reasonable solution?

I guess it could be a viable solution. If the user is setting the network capacity I would assume they know what to put there, if not and they put a wrong/bad value, they should know that it's going to screw up the estimation. The documentation should state that. If we think it's not a viable solution then we should remove the `estimatedTimeToCompletionInMinutes` from the overall proposal. But I am for taking it and documenting it properly. 

@tomncooper wdyt about the above discussion?

>   completedDataMovementPercentage: 80% [2]

Reviewer 2: You should either have the unit in the value and have users parse it or have it in the name and use an integer / double only. I'm fine with both ways, but you should pick one.

Reviewer 3: My view is ... 
For the `completedDataMovementPercentage` field it doesn't make much sense to have the `%` symbol in the value so I would be to just remove it and having `completedDataMovementPercentage: 80`.
Regarding the `estimatedTimeToCompletionInMinutes`, it could depends if we want the flexibility of showing the value in a different unit, but I don't see any value in it. I mean we could have `estimatedTimeToCompletio: 300000ms` or `estimatedTimeToCompletion: 5m` to say the same but does it make really sense? for this reason I would more be for just `estimatedTimeToCompletionInMinutes: 5`. The rebalancing is a long process and showing 1 minute or just 0 minute for a remaining time which is less than a minute could make sense.

   "minimumConcurrentPartitionMovementsPerBroker":5,
   "numFinishedPartitionMovements":0,
   "numInProgressPartitionMovements":0,
   "numPendingPartitionMovements":20,
   "numTotalPartitionMovements":20,
   "state":"INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS",
   "totalDataToMove":0,
   "triggeredSelfHealingTaskId":"",
   "triggeredTaskReason":"No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
   "triggeredUserTaskId":"0230d401-6a36-430e-9858-fac8f2edde93"
  }
  brokerLoad.json: {...} [4]
```
[1] The estimated time it will take in minutes for the rebalance to complete based on the average rate of data transfer.

[2] The percentage complete of the ongoing rebalance in the range [0-100]%

[3] The “non-verbose” JSON payload from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint.

[4] The broker load from the optimization proposal as a JSON string that already maintained in the `ConfigMap`.

### Field values per `KafkaRebalance` State

We will provide the progress information for the `KafkaRebalance` states where it is relevant:

- `ProposalReady`
- `Rebalancing`

> - `Stopped`

Reviewer 2: What will be the actual values in these states?
* Proposal ready -> 0% completion and the estimated time from once it would be approved?
* Rebalancing -> an up to date information?
* Stopped -> the last infor before it was stopped?
* Ready -> 100% and 0 minutes remaining?

Maybe you can describe it here in bullet points in a human readable form and leave the formulas below for _experts_.

Reviewer 3: +1

- `Ready`

All the information required for estimating the values of `estimatedTimeToCompletionInMinutes` and `completedDataMovementPercentage` fields can be derived from the Cruise Control server configurations and the [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) REST API endpoint.
However, the actual formulas used to produce values for these fields depend on the state of the `KafkaRebalance` resource.
Checkout the example in the [Field: executorState](#field-executorstate) section to see where the fields used in the formulas below come from.

### Field: `estimatedTimeToCompletionInMinutes`

The estimated time it will take in minutes for a rebalance to complete based on the average rate of data transfer.

> The formulas used to calculate field value per `KafkaRebalance` state:

Reviewer 2: So the formulas here ... will the CO calculate these? Or does CC calculate this and we just show the numbers?

Reviewer 3: It's the CO calculating these.

Author: Added a sentence in the section above this sentence to make this more clear

"All the information required for the Cluster Operator to estimate the values of `estimatedTimeToCompletionInMinutes` and `completedDataMovementPercentage` fields"

**Estimation for inter-broker rebalance:**

This estimate will be a theoretical minimum derived from Cruise Control capacity and throttle configurations.
This means that the cluster rebalance would take at least the estimated amount of time to complete.

$$\text{maxPartitionMovements}_{[1]} = \min(\text{numberOfBrokers} \times \text{num.concurrent.partition.movements.per.broker}),\text{max.num.cluster.partition.movements})$$

$$\text{bandwidth}_{[2]} = \min(\text{networkCapacity}, \text{replication.throttle})$$

$$\text{throughput}_{[3]} = \text{maxPartitionMovements} \times \text{bandwidth}$$

$$\text{estimatedTimeToCompletionInMinutes} = \frac{\text{dataToMoveMB}}{\text{throughput}}$$

Notes:
- [1] The maximum number of partition movements given Cruise Control partition movement capacity.

- [2] The network bandwidth given Cruise Control bandwidth throttle.

- [3] The throughput given the max allowed number of partition movements and network bandwidth.

**Estimation for intra-broker rebalance:**

It is challenging to provide an accurate estimate for intra-broker rebalances without an estimate for disk read/write throughput and getting disk throughput is non-trivial for Strimzi.

> However, by using the network bandwidth in place of the disk throughput, we can provide a rough estimate of how long the rebalance would take.

Reviewer 3: "by using the network bandwidth in place of the disk throughput" why do you think that we can do this "replacement" to get a rough estimation? I am not sure about that. I was wondering if we should just avoid this estimation if we don't have a good way for that.

Author: Assuming the disk throughput is always greater than the network bandwidth, an estimate using the network bandwidth could serve as an upperbound (theoretical maximum) of how long an intra-broker rebalance would take. e.g the rebalance won't take longer than this. However, this contradicts the definition provided by the inter-broker balance and may cause confusion. We could simply set the value to `N/A` for now to avoid confusion/inaccuracy. 

Would you mind if we left this estimate out for `intra-broker` balances @tomncooper?

Reviewer: I thought that is what we were going to do? Basically we don't include the intra estimate as we can't reliably calculate it. So the time to completion is always a theoretical minimum, it WILL take longer than this.

Reviewer 3: Yes I would leave this estimation out imho.

Author: Updated proposal to explain this and suggest that we set it to "N/A"

$$\text{estimatedDiskThroughput} = \text{???}_{[2]}$$

$$\text{bandwidth}_{[3]} = \min(\text{networkCapacity}, \text{replication.throttle})$$

$$\text{throughput}_{[4]} = \text{maxPartitionMovements} \times 
\begin{cases}
  \text{estimatedDiskThroughput}, & \text{if available} \\
  \text{bandwidth}, & \text{otherwise}
\end{cases}
$$

$$\text{estimatedTimeToCompletionInMinutes} = \frac{\text{intraBrokerDataToMoveMB}}{\text{throughput}}$$

Notes
- [1] The maximum number of partition movements given Cruise Control partition movement capacity
- [2] We don't have a method of determining disk throughput.
- [3] The network bandwidth given Cruise Control bandwidth throttle.
- [4] The throughput given the max allowed number of partition movements and disk throughput.
 
#### State: `Rebalancing`

$$
\text{rate} = \frac{\text{finishedDataMovement}\_{[1]}}{\text{taskTriggerTime}\_{[2]} - \text{currentTime}}
$$

$$
\text{estimatedTimeToCompletionInMinutes} = \frac{\text{totalDataToMove}\_{[3]} - \text{finishedDataMovement}\_{[1]}}{\text{rate}}
$$

Notes

> - [1] The number of megabytes already moved by rebalance, provided by [/kafkacruisecontrol/state?substates=executor](#field-executorstate) REST API endpoint.

Reviewer 3: Can we put the specific field we are using from the returned JSON?

- [2] The time when the rebalance task was started, extracted from `triggeredTaskReason` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate) for that task.

> - [3] The total number of megabytes planned to be moved for rebalance, provided from json payload of the [/kafkacruisecontrol/state?substates=executor](#field-executorstate) REST API endpoint.

Reviewer 3: Can we put the specific field we are using from the returned JSON?

#### State: `Stopped`

$$
\text{estimatedTimeToCompletionInMinutes} = \text{N/A}
$$

Once a rebalance has been stopped, it cannot be resumed. 
Therefore, there is no `estimatedTimeToCompletionInMinutes` for a stopped rebalance. We set the field to `N/A` to emphasize this. 
To move from the `Stopped` state, a user must refresh the `KafkaRebalance` resource, the `rebalanceProgressConfigMap` will then be updated by the operator upon the next state change.

#### State: `Ready`

$$
\text{estimatedTimeToCompletionInMinutes} = 0
$$

The rebalance is complete so we hardcode the value to `0`
This emphasizes that the rebalance is complete and helps clear up ambiguity surrounding what the `Ready` state means in the `KafkaRebalance` resource.

### Field: `completedDataMovementPercentage`

The percentage of the data movement of the partition rebalance that is completed.

> The formulas used to calculate field value per  `KafkaRebalance` state:

Reviewer 5: ```suggestion
The formulas used to calculate the field value differ for each applicable `KafkaRebalance` state:
```

#### State: `ProposalReady`

$$
\text{completedDataMovementPercentage} = 0\%
$$

The rebalance has not started yet so no data has been moved, so we hardcode the value to `0`.
This emphasizes that the rebalance has not started and helps clear up ambiguity surrounding what the `optimizationResult` field means in the `KafkaRebalance` resource.

#### State: `Rebalancing`

$$
\text{completedDataMovementPercentage} = (\frac{\text{finishedDataMovement}\_{[1]}}{\text{totalDataToMove}\_{[2]}} \times 100)\%
$$

Notes
 - [1] The number of megabytes already moved by rebalance, provided by [/kafkacruisecontrol/state?substates=executor](#field-executorstate) REST API endpoint.

>  - [2] The total number of megabytes planned to be moved for rebalance, provided by [/kafkacruisecontrol/state?substates=executor](#field-executorstate) REST API endpoint.

Reviewer 3: Can we put the specific fields we are using from the returned JSON for both the above?

}
```

The formulas used to calculate field value per  `KafkaRebalance`  state:

#### State: `Ready`

The rebalance has not started yet so no data surrounding partition movement is available yet, so we set the field to an empty JSON object.

#### State: `Rebalancing`

The field will be assigned the "non-verbose" JSON payload of the [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint of the Cruise Control REST API.

#### State: `Stopped`

Once a rebalance has been stopped, it cannot be resumed.
However, the `executorState` information from when before the rebalance was stopped may still be of value to users.
Therefore, we reuse the same value of `executorState` from the previous update.

#### State: `Ready`

The rebalance is complete in this state so we hardcode the value to an empty JSON object.

### Progress Update Cadence

For ease of implementation and minimizing the load on the CruiseControl REST API server, the operator will only query the `/kafkacruisecontrol/state?substates=executor` endpoint and update the `ConfigMap` upon `KafkaRebalance` resource reconciliation.

> In the event that Cruise Control runs into an error when rebalancing, the operator will transition the `KafkaRebalance` resource to the `NotReady` state, remove the `progress` section, and delete the progress `ConfigMap`.

Reviewer 3: "delete the progress `ConfigMap`" or "delete the progress section of the `ConfigMap`"?
Remember we are using the same `ConfigMap` for two purposes (also storing the before/after load data).

Reviewer: Is this consistent with what currently happens if we fail to get a response from CC in a single reconciliation? Does the CC client retry? Not sure setting the KR CR to `NotReady` for what could be a simple network blip is good UX.

Author: > delete the progress ConfigMap" or "delete the progress section of the ConfigMap"?
Remember we are using the same ConfigMap for two purposes (also storing the before/after load data).

I am going to update the behavior described here to retain the progress information and `ConfigMap` when the `KafkaRebalance` resource moves to the `NotReady` state. As discussed in a previous thread with Paul, the progress information in the `NotReady` state may just be as useful for debugging as it is in the `Stopped`.

> Is this consistent with what currently happens if we fail to get a response from CC in a single reconciliation? Does the CC client retry? Not sure setting the KR CR to NotReady for what could be a simple network blip is good UX.

This line was intended to describe how the progress information will be updated when CC server returns "CompletedWithError" status for a task. From what I understood when writing this, this was the only situation where the KR resource was moved to the `NotReady` state. But looking closer at the code, it appears I am wrong.

It appears the CO will move the KR resource to the  `NotReady` state when it fails to get a response from the CC server, it also looks like the CO CC client code does not retry when failing to get a response (unless I am missing some retry logic in the code). This means that if the CO fails to get a response from CC server it will set the KR CR is set to `Not Ready`. I thought this only happened when there was an "CompleteWithError" response returned by the CC server, not when there was a failed HTTP request. 

The proposal suggests attempting to retrieve the executor status but whether the retrieval succeeds or fails has no affect on the state of the `KafkaRebalance` resource.

Of course, the CC client, wherever it is used, should and will be implemented to retry

In the event that the Cruise Control REST API returns an error or fails to respond to the operator when querying the `/kafkacruisecontrol/state?substates=executor` endpoint during a rebalance, the operator will add a "Warning" condition with error message from the failed REST API call but leave the existing progress section and progress `ConfigMap` as is.

When Cruise Control state retrieval fails, the `KafkaRebalance` resource will be updated as follows:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  ...
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"

>     type: Warning

Reviewer 5: Should we call out this property to highlight and distinguish between `NotReady` and (new?) `Warning`?

Author: The type of condition in the status could be any one of the `KafkaRebalance` states, it could also be `Warning`. Why would the `NotReady` type be tied/associated to the `Warning` type?

Reviewer 2: @kyguy I think this is a good point from Paul that confused me as well -> you should have one of the regular states included in the example if they would accompany the warning.

Author: Sorry Paul, I misunderstood your suggestion, I'll add a `KafkaRebalance` state condition to the example to draw a distinguish between the `KafkaRebalance` state condition and the `Warning` condition.

    reason: CruiseControlRestException
    message: <error_message> [1]  
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress:
    rebalanceProgressConfigMap: my-rebalance-progress
```

> [1] Error message from failed Cruise Control REST API call

Reviewer 5: ```suggestion
[1] Error message from failed Cruise Control REST API call.
```

### Accessing progress fields using Kubernetes CLI

The progress information will be stored in a `ConfigMap` with the same name as the `KafkaRebalance` resource.

> Using the name of the ConfigMap we can view its data from the command line using the Kubernetes CLI.

Reviewer 5: ```suggestion
Using the name of the `ConfigMap`, we can view its data from the command line using the Kubernetes CLI.
```

Example accessing `estimatedTimeToCompletionInMinutes` field.
```shell
kubectl get configmaps <my_rebalance_configmap_name> -o json | jq '.["data"]["estimatedTimeToCompletionInMinutes"]'
```

Example accessing `completedDataMovementPercentage` field.
```shell
kubectl get configmaps <my_rebalance_configmap_name> -o json | jq '.["data"]["completedDataMovementPercentage"]'
```

Example accessing `executorState` field.
```shell
kubectl get configmaps <my_rebalance_configmap_name> -o json | jq '.["data"]["executorState"] | fromjson | .'
```

### Rejected Alternatives

> #### Maintaining progress fields in KafkaRebalance resource status

Reviewer 5: ```suggestion
#### Maintaining progress fields in `KafkaRebalance` resource status
```

   "numFinishedPartitionMovements":0,
   "numInProgressPartitionMovements":0,
   "numPendingPartitionMovements":20,
   "numTotalPartitionMovements":20,
   "state":"INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS",
   "totalDataToMove":0,
   "triggeredSelfHealingTaskId":"",
   "triggeredTaskReason":"No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
   "triggeredUserTaskId":"0230d401-6a36-430e-9858-fac8f2edde93"
  }
  brokerLoad.json: {...} [4]
```
[1] The estimated time it will take in minutes for the rebalance to complete based on the average rate of data transfer.

[2] The percentage of byte movement of the partition balance that is complete as a rounded down integer in the range [0-100].

[3] The “non-verbose” JSON payload from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint.

[4] The broker load from the optimization proposal as a JSON string that already maintained in the `ConfigMap`.

### Field values per `KafkaRebalance` State

We will provide the progress information for the `KafkaRebalance` states where it is relevant:

> | State           | `estimatedTimeToCompletionInMinutes`                                   | `completedByteMovementPercentage`                                                                                                    | `executorState`                                                                                                                                                                                      |

Reviewer 3: I would use "State/Field" in the first cell to make it clear, rows are states and columns are fields.

This `ConfigMap` will also be enhanced to contain information related to an ongoing partition rebalance.

The `progress` section of the `KafkaRebalance` resource will look like the following:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: ProposalReady | Rebalancing | Stopped | Ready
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress: [1]

>     rebalanceProgressConfigMap: my-rebalance [2]

Reviewer 4: What's the value of having this nested structure compared to just `.status.rebalanceProgressConfigMap`?

Author: To further organize/distinguish the progress information from the other status fields.

```
[1] The `progress` section will be visible during the `ProposalReady`, `Rebalancing`, `Stopped`, `NotReady` and `Ready` states.

[2] The `ConfigMap` containing information related to the ongoing partition rebalance.

In the `ConfigMap`, we will add the following fields:
- **estimatedTimeToCompletionInMinutes**: The estimated amount time it will take in minutes until the partition rebalance is complete.
- **completedByteMovementPercentage**: The percentage of the byte movement of the partition rebalance that is completed as a rounded down integer value in the range [0-100].

> - **executorState**: The “non-verbose” JSON payload from the [`/kafkacruisecontrol/state?substates=executor`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint, providing details about the executor's current status, including partition movement progress, concurrency limits, and total data to move.

Reviewer 3: Following the discussion here https://github.com/strimzi/strimzi-kafka-operator/issues/11133 I have actually one suggestion: could we call this as `executorState.json` to be consistent with other fields containing JSON stuff?

Since the progress information is constant, we can safely add it to the existing `ConfigMap` maintained for and tied to the `KafkaRebalance` resource.

> This keeps `KafkaRebalance` information organized in one place, simplifies the proposal implementation, and has insignificant impact on the storage of the `ConfigMap`.

Reviewer 4: Are we sure that mixing unrelated information in the same CM is actually a good idea? What's the complication of having a dedicated progress CM?

Author: > Are we sure that mixing unrelated information in the same CM is actually a good idea?

Is the load and progress information really unrelated? Couldn't we as easily think of the information as being related to a specific rebalance? 

> What's the complication of having a dedicated progress CM?

We were debating whether or not to have a dedicated progress CM. One of the arguments against having a dedicated CM was that it would require extra API calls and code wrangling in the `KafkaRebalanceAssemblyOperator` class all while we already maintain `ConfigMap` for a `KafkaRebalance` resource with plenty of space for the progress information. We would also need to change the name of the existing `ConfigMap` to differentiate it from a new one since the existing name matches the name `KafkaRebalance` resource.

Reviewer 3: Yeah having just one simplifies the implementation as well. At the same time one place to look at for the user. Taking into account the amount the information we are adding, it sounded not taking any advantage from having two CMs.

Reviewer 4: Thanks for the clarifications. I can live with the single CM, but then I would prefer a single status field with a meaningful name, deprecating the old one. Have you considered this alternative design?

Reviewer 4: Thanks for considering this change. As always, naming things is hard. The following schema should give more flexibility:

```sh
# this proposal
status:
  loadAndProgressConfigMaps:
    - my-rebalance-progress

# glimpse into a possible future
status:
  loadAndProgressConfigMaps:
    - my-rebalance-load
    - my-rebalance-progress
    - my-rebalance-progress-verbose
```

Reviewer: With this proposal we have 2 fields (load and progress) referencing the same CM. If that CM every becomes to large then we could have each reference its own CM. If either of them become too big we probably need a whole different way of communicating that information to the user. 

So I think it is better to stick with the current plan than deprecate an existing field and add 2 more.

Reviewer 3: Yeah I agree with Tom subscribing what he said. The proposal from Fede looks to be more "complicated" imho.

Author: Unless there are any objections, let's stick with the current plan, we should have enough flexibility for now and the future.

However, the actual formulas used to produce values for these fields depend on the state of the `KafkaRebalance` resource.
Checkout the example in the [Field: executorState.json](#field-executorstate.json) section to see where the fields used in the formulas below come from.

### Field: `estimatedTimeToCompletionInMinutes`

The estimated time it will take in minutes for a rebalance to complete.

The formulas used to calculate the field value differ for each applicable `KafkaRebalance` state:

#### State: `ProposalReady`

Since the rebalance has not started at this point, we do not have the data needed to easily provide an accurate value for this field.
Therefore, we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.

#### State: `Rebalancing`

The estimated time it will take for the ongoing rebalance to complete based on the average rate of data transfer.

$$
r = \frac{D_{moved}}{T_{\text{current}} - T_{\text{start}}}
$$

$$

> ETC = \frac{D_{total} - D_{moved}}{r \cdot 60000}

Author: Just added a note now.

$$

**Notes:**

> - $r$: The average rate of data transfer in megabytes per second, calculated as the ratio of finished data movement to the elapsed time.

Author: Argh yes, I missed this in my initial attempt to fix the units

- $D_{moved}$: The number of megabytes already moved by the rebalance, provided by `finishedDataMovement` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
- $T_{\text{current}}$: The current time when the estimate is being calculated, expressed as the number of milliseconds since the Unix epoch.
- $T_{\text{start}}$: The time when the rebalance task was triggered, expressed as the number of milliseconds since the Unix epoch, extracted from the `triggeredTaskReason` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
- $D_{total}$: The total number of megabytes planned to be moved for the rebalance, provided by `totalDataToMove` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
- $ETC$: The estimated time to completion in minutes based on the average rate of data transfer, the value of the `estimatedTimeToCompletionInMinutes` field.

#### State: `Stopped`

Once a rebalance has been stopped, it cannot be resumed. 

> Therefore, there is no `estimatedTimeToCompletionInMinutes` for a stopped rebalance so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.

Author: Yes, it won't be there

To move from the `Stopped` state, a user must refresh the `KafkaRebalance` resource, the `rebalanceProgressConfigMap` will then be updated by the operator upon the next state change.

#### State: `NotReady`

Once a rebalance has been interrupted or completed with errors, it cannot be resumed.
Therefore, there is no `estimatedTimeToCompletionInMinutes` for the rebalance so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
To move from the `NotReady` state, a user must refresh the `KafkaRebalance` resource, the `rebalanceProgressConfigMap` will then be updated by the operator upon the next state change.

#### State: `Ready`

> The rebalance is complete so we hardcode the value to `0`

Reviewer 4: What's the value of setting `estimatedTimeToCompletionInMinutes: 0` once you have `completedByteMovementPercentage: 100`? Can't we simply remove this field like in the other cases? It is also semantically incorrect, as this is not an estimation.

Author: > What's the value of setting `estimatedTimeToCompletionInMinutes`: 0 once you have `completedByteMovementPercentage`: 100

For someone tracking the `estimatedTimeToCompletionInMinutes` field alone, this is still valuable to know when a rebalance is complete.

> Can't we simply remove this field like in the other cases? 

Given we remove this field in other `KafkaRebalance` states where a rebalance cannot complete, this information is especially useful to emphasize that a rebalance is complete.

Reviewer 4: Fair enough, thanks.

This emphasizes that the rebalance is complete and helps clear up ambiguity surrounding what the `Ready` state means in the `KafkaRebalance` resource.

> ### Field: `completedByteMovementPercentage`

Reviewer: The "byte" part of this smells wrong, why not "data"?

Author: It was raised in an earlier comment that "data" movement could be misinterpreted as partition movement instead of the byte movement. Naming this field with "byte" removes any ambiguity as to what is being measured

Reviewer: Not sure it would be easy to mix up data and partitions personally, but ok. I still think it would be better to have `completedDataMovementPercentage` and `completedPartitionMovementPercentage` then you can't mix them up.

Author: >  I still think it would be better to have `completedDataMovementPercentage` and `completedPartitionMovementPercentage` then you can't mix them up.

I was worried that information from these two fields would be too similar, therefore, I was hoping to only supply one of them. However, I want the field name(s) to be as clear as possible to everyone. I am open to including both, it was something which @katheris raised in an earlier review too.

I would be interested in what @scholzj thinks of this.

Author: Discussed this with Jakub on a call today, even if we were to include `completedPartitionMovementPercentage` the field in the proposal, the "Data" in `completedDataMovementPercentage` field still seems that it could be potentially ambiguous. 

If using the field name `completedByteMovementPercentage` is not a blocker for anyone, let's stick with this for now. We can always add the `completedPartitionMovementPercentage` field in a later proposal too!

CC @katheris would still like your opinion on this too ^^^

The percentage of the byte movement of the partition rebalance that is completed as a rounded down integer in the range [0-100].

The formulas used to calculate field value per  `KafkaRebalance` state:

#### State: `ProposalReady`

The rebalance has not started yet so no data has been moved, so we hardcode the value to `0`.
This emphasizes that the rebalance has not started and helps clear up ambiguity surrounding what the `optimizationResult` field means in the `KafkaRebalance` resource.

#### State: `Rebalancing`

> The percentage of the byte movement of the partition rebalance that is completed as a rounded down integer in the range [0-100].

Reviewer: What is the motivation for rounding down rather than just rounding to the nearest integer?

Reviewer: So long as that is documented somewhere it should be fine. The only issue I can see is that for a very large rebalance you might end up waiting for a while for the percentage to move off 0%.

$$
\text{DMP} = \left\lfloor \frac{D_{moved}}{D_{total}} \times 100 \right\rfloor
$$

**Notes:**

> - $DMP$: The percentage of byte data that has been moved as a rounded down integer in the range [0-100], the value of the `completedByteMovementPercentage` field.

Reviewer: "byte data" sounds weird? You can probably just drop it an use data instead?

### Progress Update Cadence

For ease of implementation and minimizing the load on the CruiseControl REST API server, the operator will only query the `/kafkacruisecontrol/state?substates=executor` endpoint and update the `ConfigMap` upon `KafkaRebalance` resource reconciliation.

In the event that Cruise Control runs into an error when rebalancing, when operator transitions the `KafkaRebalance` resource to the `NotReady` state, it will leave the `progress` section and `ConfigMap` as is.
In the event that the Cruise Control REST API returns an error or fails to respond to the operator when querying the `/kafkacruisecontrol/state?substates=executor` endpoint during a rebalance, the operator will add a "Warning" condition with error message from the failed REST API call but leave the existing progress section and progress `ConfigMap` as is.

When Cruise Control state retrieval fails, the `KafkaRebalance` resource will be updated as follows:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  ...
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: Rebalancing

>   - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"

Reviewer 6: This might be unlikely case so maybe does not matter, but if Cruise Control REST API returns an error in several reconciliations, would `lastTransitionTime` be updated in  every reconciliation or only if the warning message/state changed? I was wondering how this reflects on the progress field to indicate how out of date `estimatedTimeToCompletionInMinutes` might be.

Author: >  would lastTransitionTime be updated in every reconciliation or only if the warning message/state changed

This is a good point. In the event Cruise Control REST API returns an error for several reconciliations, the condition would be updated only when the warning message/state changed. This would highlight  error ongoing errors would better indicate how stale `estimatedTimeToCompletionInMinutes` is.

Let me add a line to clarify this.

Author: @tinaselenge Does the added line make this more clear?

    status: "True"
    type: Warning
    reason: CruiseControlRestException
    message: <error_message> [1]  
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress:
    rebalanceProgressConfigMap: my-rebalance
```
[1] Error message from failed Cruise Control REST API call.

### Accessing progress fields using Kubernetes CLI

The progress information will be stored in a `ConfigMap` with the same name as the `KafkaRebalance` resource.
Using the name of the `ConfigMap` we can view its data from the command line using the Kubernetes CLI.

Example accessing `estimatedTimeToCompletionInMinutes` field.
```shell

> kubectl get configmaps <my_rebalance_configmap_name> -o json | jq '.["data"]["estimatedTimeToCompletionInMinutes"]'

Reviewer: What happens when you run this command and the field is "ommited" as you said above for some of the states. It will probably fail, but you should probably mention that.

Author: Fair point, added a note for this.

  completedByteMovementPercentage: 80 [2]
  executorState.json: [3]
  {
   "abortingPartitions":0,
   "averageConcurrentPartitionMovementsPerBroker":5,
   "finishedDataMovement":0,
   "maximumConcurrentPartitionMovementsPerBroker":5,
   "minimumConcurrentPartitionMovementsPerBroker":5,
   "numFinishedPartitionMovements":0,
   "numInProgressPartitionMovements":0,
   "numPendingPartitionMovements":20,
   "numTotalPartitionMovements":20,
   "state":"INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS",
   "totalDataToMove":0,
   "triggeredSelfHealingTaskId":"",
   "triggeredTaskReason":"No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
   "triggeredUserTaskId":"0230d401-6a36-430e-9858-fac8f2edde93"
  }
  brokerLoad.json: {...} [4]
```
[1] The estimated time it will take in minutes for the rebalance to complete based on the average rate of data transfer.

[2] The percentage of byte movement of the partition balance that is complete as a rounded down integer in the range [0-100].

[3] The “non-verbose” JSON payload from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint.

> [4] The broker load from the optimization proposal as a JSON string that already maintained in the `ConfigMap`.

Author: The broker load information is available for all these states but I don't think we should not include it in the table since it isn't progress information that we are adding as part of this proposal. That being said, I think we should add a table like this in the documentation including the broker load information for the implementation of this proposal.

### Future Improvements

#### Providing theoretical minimum estimate for `estimatedTimeToCompletionInMinutes` field for `ProposalReady` state

In addition to having an estimate of how long a rebalance will take _during_ a rebalance, it would also be useful to have a theoretical minimum estimate of how long a rebalance would take _before_ a rebalance has started.
However, providing such estimation is non-trivial.
It would be better addressed in its own proposal where we can have deeper discussion and analysis.

Leveraging the Cruise Control configurations and user-provided network capacity settings, we could provide a rough estimate for the `estimatedTimeToCompletion` field but there are a couple of challenges.
For inter-broker rebalances, the accuracy of the estimation is dependent on the accuracy of the user-defined network capacity settings.
For intra-broker rebalances, the accuracy of the estimation is dependent on broker disk read/write bandwidth.
These are only a couple of the challenges and complications of providing such an estimate, in a future proposal we can discuss in more detail.

> ** Rough estimation for inter-broker rebalances:**

Reviewer 5: ```suggestion
**Rough estimation for inter-broker rebalances:**
```

We could only provide an accurate estimate for inter-broker rebalances with an accurate estimate of the network capacity of the brokers.
Therefore, the calculation would be dependent on how well users configure their network capacity settings.

$$
ETC = \frac{D_{total} \cdot 1024 \cdot 1024}{\min\left(NB, R_t\right) \cdot 60}
$$

Notes:
- $ETC$: Estimated time to completion in minutes, the value of the `estimatedTimeToCompletionInMinutes` field.
- $D_{total}$: The total number of megabytes planned to be moved for the rebalance, provided by `dataToMoveMB` field in `status.optimizationResult` section of `KafkaRebalance` resource.
- $NB$: The average network bandwidth per broker in the cluster in bytes per second, derived from the `spec.cruiseControl.brokerCapacity.inboundNetwork` and `spec.cruiseControl.brokerCapacity.outboundNetwork` fields in the `Kafka` custom resource.
- $R_t$: The replication throttle applied to replicas being moved in and out per broker in bytes per second, provided by Cruise Control  [`default.replication.throttle`](https://github.com/linkedin/cruise-control/wiki/Configurations) configuration.

> ** Rough estimation for intra-broker rebalances:**

Reviewer 5: ```suggestion
**Rough estimation for intra-broker rebalances:**
```

### Field values per `KafkaRebalance` State

We will provide the progress information for the `KafkaRebalance` states where it is relevant:

- `ProposalReady`:
  - `completedByteMovementPercentage`: 0
- `Rebalancing`:
  - `estimatedTimeToCompletionInMinutes`: The estimated time it will take for the ongoing rebalance to complete.
  - `completedByteMovementPercentage`: The percentage of byte movement of the ongoing partition rebalance that is complete as a rounded down integer in the range [0-100].
  - `executorState`: JSON object from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint of Cruise Control REST API.
- `Stopped`:
  - `completedByteMovementPercentage`: The value of `completedByteMovementPercentage` from previous update.
  - `executorState`: JSON object from previous update.
- `NotReady`:
  - `completedByteMovementPercentage`: The value of `completedByteMovementPercentage` from previous update.
  - `executorState`: JSON object from previous update.
- `Ready`:
  - `estimatedTimeToCompletionInMinutes`: 0

>   - `completedByteMovementPercentage`: 100

Author: Great idea,this will look a lot neater, I transposed the suggested table but it looks just as neat, much clearer than having the list

When Cruise Control state retrieval fails, the `KafkaRebalance` resource will be updated as follows:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  ...
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: Rebalancing
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: Warning
    reason: CruiseControlRestException
    message: <error_message> [1]  
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress:

>     rebalanceProgressConfigMap: my-rebalance-progress

Reviewer 3: this can't be `my-rebalance-progress` if we have `afterBeforeLoadConfigMap: my-rebalance` ... are we going to use the same ConfigMap for both, right? So the naming should be consistent in the example.

Author: Thank you for the catch, just fixed

This `ConfigMap` will also be enhanced to contain information related to an ongoing partition rebalance.

The `progress` section of the `KafkaRebalance` resource will look like the following:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: ProposalReady | Rebalancing | Stopped | Ready [1]
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress: [1]
    rebalanceProgressConfigMap: my-rebalance [2]
```

> [1] The `progress` section will be visible during the `ProposalReady`, `Rebalancing`, `Stopped`, `NotReady` and `Ready` states.

Reviewer 3: based on the decision we'll make, I guess we should remove `ProposalReady` from here and in any other place where it's mentioned.

Author: Even if we decide to not provide a forecast value for the `estimatedTimeToCompletionInMinutes` field in the `ProposalReady` state, would there be any value still supporting this state for showing `completedDataMovementPercentage` = 0 to emphasize that the rebalance hasn't started? Of course it this state, `estimatedTimeToCompletionInMinutes` would be assigned `-1` or whatever value we decide to represent that it is not available.  What do you think?

# Adding progress updates for Cruise Control rebalances

This proposal introduces a new feature to monitor the progression of an ongoing partition rebalance executed by a Strimzi-managed Cruise Control instance via a `KafkaRebalance` custom resource. 

## Current Situation

At this time, Strimzi users are able to execute partition rebalances via `KafkaRebalance` custom resources but can only monitor the progression of those partition rebalances in two ways:

- Manually querying the Cruise Control REST API endpoint directly. 
- Inspecting the logs of the Cruise Control instance. 

Unfortunately, neither of these methods are particularly user-friendly.

## Motivation

Information surrounding the progress of an executing partition rebalance is useful for planning future cluster operations. 
Knowing things like how much time an ongoing partition rebalance has left to take and how much data an ongoing partition rebalance has left to transfer helps users understand the cost of an ongoing partition rebalance.

> This information helps users decide whether they should continue or cancel an ongoing rebalance, and know when future operations will be able to be safely executed.

Author: Maybe I am can use a better term here than "cluster operations", what I was thinking of was node maintenance, planned and unplanned, like if SREs need to address hardware issues and cannot wait for a partition rebalance to run to completion.

> This information helps users decide whether they should continue or cancel an ongoing rebalance, and know when future operations will be able to be safely executed.

Reviewer 2: I'm not convinced how valid this logic really is. From my experience, these will be very often dealt with by different teams. But for the sake of this proposal, I think we should have:
* More detail on what `cluster operations` mean here ... for example as suggested, worker node maintenance etc.
* More careful wording about what can be `safely executed`. While you probably want to have the rebalance finished before you start doing some draining, it still needs to be safe regardless because these operations happen in most cases automatically.

> This information helps users decide whether they should continue or cancel an ongoing rebalance, and know when future operations will be able to be safely executed.

Author: You are right, `safely executed` isn't the right phrase for this as the rebalance operations are generally safe. A better explanation for the motivation for this feature is that it allows more optimal scheduling of these operations. Just updated the examples and explanation to reflect this, let me know if it is suitable or if it needs better phrasing!

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
spec: {}
status:
  conditions:
  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
    status: "True"
    type: ProposalReady | Rebalancing | Stopped | Ready [1]
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 0
    numLeaderMovements: 16
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress: [1]
    rebalanceProgressConfigMap: my-rebalance [2]
```
[1] The `progress` section will be visible during the `ProposalReady`, `Rebalancing`, `Stopped`, `NotReady` and `Ready` states.

[2] The `ConfigMap` containing information related to the ongoing partition rebalance.

In the `ConfigMap`, we will add the following fields:

> - **estimatedTimeToCompletionInMinutes**: The estimated amount time it will take in minutes until partition rebalance is complete.

Reviewer 2: Actually, even with the throughput set properly, it seems you would get a completely random value if you don't know how much of the throughput is used for the regular communication.

> - **estimatedTimeToCompletionInMinutes**: The estimated amount time it will take in minutes until partition rebalance is complete.

Author: > Following on the previous discussion ... I think this should be removed as long as we do not have a proper solution for the network capacity and thus this number cannot be reliably provided in most cases. And right now, we do not seem to have any answers for the network capacity.

I assume this comment is for the `estimatedTimeToCompletionInMinutes` in the `ProposalReady` state, this is the only state where the calculation relies on network capacity settings. 

Yes, this calculation depends on a valid network capacity value to be provided.

> - **completedDataMovementPercentage**: The percentage of the data movement of the partition rebalance that is completed e.g. values in the range [0-100].

Reviewer 2: What is data movement? Is that percentage of bytes? Or percentage of partition-replicass being moved? I.e., if I have proposal moving 20 partition replicas with total amount of 1Gi of data and right now we have moved 16 partitions with 200Mi of data ... will the percentage be 80% or 20%? I think the name should make this clear.

Also, will it be an integer? Or will it be more precise? If it is integer, how will it be rounded?

Author: The data movement here is the percentage of bytes, maybe we could rename to `completedByteMovementPercentage` to make this more clear, would that be suitable?

> Also, will it be an integer? Or will it be more precise? If it is integer, how will it be rounded?

It will be an integer and rounded down to be more conservative. I'll update the proposal with this information.

Reviewer 2: `completedByteMovementPercentage` sounds good as well as integrer rounded down. Please make sure it is included int he proposal.

Author: Updated

[3] The “non-verbose” JSON payload from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint.

[4] The broker load from the optimization proposal as a JSON string that already maintained in the `ConfigMap`.

### Field values per `KafkaRebalance` State

We will provide the progress information for the `KafkaRebalance` states where it is relevant:

- `ProposalReady`:
  - `estimatedTimeToCompletion`: The theoretical minimum estimated time it would take the rebalance to complete if executed.
  - `completedDataMovementPercentage`: 0
  - `executorState`: Empty JSON object.
- `Rebalancing`
  - `estimatedTimeToCompletion`: The estimated time it will take for the ongoing rebalance to complete.
  - `completedDataMovementPercentage`: The percentage of data movement of the ongoing partition rebalance that is complete.
  - `executorState`: JSON object from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint of Cruise Control REST API.
- `Stopped`

>   - `estimatedTimeToCompletion`: N/A

Reviewer 2: Here and below ... what exactly does `N/A` mean? Null value? Empty string? The field will not be set? `N/A` as a String?

Author: It will be a String "N/A" to represent "not applicable" or "not available".

Reviewer 2: Is that really the right way how to do this? Sounds like it would be hard to parse in client applications.

Reviewer 3: Yeah, we are discussing the same in another comment here https://github.com/strimzi/proposals/pull/140#discussion_r1925620859

  - `executorState`: JSON object from previous update.
- `NotReady`
  - `estimatedTimeToCompletion`: N/A
  - `completedDataMovementPercentage`: The value of `completedDataMovementPercentage` from previous update.
  - `executorState`: JSON object from previous update.
- `Ready`:
  - `estimatedTimeToCompletion`: 0
  - `completedDataMovementPercentage`: 100
  - `executorState`: Empty JSON object.

All the information required for the Cluster Operator to estimate the values of `estimatedTimeToCompletionInMinutes` and `completedDataMovementPercentage` fields can be derived from the Cruise Control server configurations and the [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) REST API endpoint.
However, the actual formulas used to produce values for these fields depend on the state of the `KafkaRebalance` resource.
Checkout the example in the [Field: executorState](#field-executorstate) section to see where the fields used in the formulas below come from.

### Field: `estimatedTimeToCompletionInMinutes`

The estimated time it will take in minutes for a rebalance to complete.

> The formulas used to calculate the field value differ for each applicable `KafkaRebalance` state:

Reviewer 2: The data movement percentage calculations seem pretty simple. But these seem quite complex. Is it really Strimzi's business to have some formulas to calculate things like this? Shouldn't that be provided by CC or calculated by the users?

Reviewer 3: > Shouldn't that be provided by CC

Despite the complexity you are mentioning, I don't think we are getting this info from CC any time soon even if we try to propose something like this. Why do you see this as something that the operator should not do based on the data retrieved by CC?

> or calculated by the users?

What do you mean? Manually? We should provide some progress information and the user should have their own logic to process them and make the numbers?

Reviewer 2: > We should provide some progress information and the user should have their own logic to process them and make the numbers?

Yes. The concern is that the various formular presented here - aside from the complexity - would never be good enough for majority of users and you will have forever discussions how to _improve them_.

That said, I think this is valid mainly for the time estimation, where I think there are also other more serious concerns such as the dependency on network bandwidth.

Author: > Yes. The concern is that the various formular presented here - aside from the complexity - would never be good enough for majority of users and you will have forever discussions how to improve them.

I agree that the formula for the `ProposalReady` state is complex 

> That said, I think this is valid mainly for the time estimation, where I think there are also other more serious concerns such as the dependency on network bandwidth.

The `estimatedTimeToCompletionInMinutes` field only requires the network bandwidth for the `ProposalReady` state. As mentioned, to make this on the other thread, we can estimate it when users provide an explict network capacity setting.

Author: >  Yes. The concern is that the various formular presented here - aside from the complexity - would never be good enough for majority of users and you will have forever discussions how to improve them.

I understand the concern about the complexity of this formula and the potential for ongoing discussions on improving them. However, for the users it would be good enough for now it would be pretty useful! We can certainly iterate and refine the approach based on user feedback to ensure it's practical and is useful for as many people as possible. Why not start here and adjust it as we learn more about how it’s used in real-world scenarios? 

> That said, I think this is valid mainly for the time estimation, where I think there are also other more serious concerns such as the dependency on network bandwidth.

The value for the `estimatedTimeToCompletionInMinutes` field is only dependent on network bandwidth for the `ProposalReady` state. To prevent displaying inaccurate estimates for this field in this state, we could provide a value only when the user has explicitly set their network capacity settings. This ensures that the value could only be incorrect if the user provided an inaccurate network capacity value.

It is important that users configure these capacity settings anyway if they want accurate partition balances based on resource usage!

Reviewer 2: > The value for the estimatedTimeToCompletionInMinutes field is only dependent on network bandwidth for the ProposalReady state. To prevent displaying inaccurate estimates for this field in this state, we could provide a value only when the user has explicitly set their network capacity settings. This ensures that the value could only be incorrect if the user provided an inaccurate network capacity value.

It would be fine with me if this is shown only conditionally based on the network throughput being set by the users. But I think it is set in completely different places, so I think it might be technically challenging to do that.

#### State: `ProposalReady`

**Estimation for inter-broker rebalance:**

This estimate will be a theoretical minimum derived from Cruise Control capacity and throttle configurations.
This means that the cluster rebalance would take at least the estimated amount of time to complete.

> $$\text{maxPartitionMovements}_{[1]} = \min(\text{numberOfBrokers} \times \text{num.concurrent.partition.movements.per.broker}),\text{max.num.cluster.partition.movements})$$

Reviewer 7: I can't follow this calculation, is it calculating the maximum number of non-concurrent partition movements? I can't work out why the number of brokers is needed. Isn't the worst case scenario that all movements have to happen from a single broker, so should it be `max.num.cluster.partition.movements/num.concurrent.partition.movements.per.broker`? Also it looks like you are missing a bracket somewhere, possible at the beginning of `numberOfBrokers x num.concurrent.partition.movements.per.broker`?

Author: @katheris There were unnecessary calculations in the originally provided formula, they have been fixed/simplified. This should now be addressed!

$$\text{bandwidth}_{[2]} = \min(\text{networkCapacity}, \text{replication.throttle})$$

$$\text{throughput}_{[3]} = \text{maxPartitionMovements} \times \text{bandwidth}$$

$$\text{estimatedTimeToCompletionInMinutes} = \frac{\text{dataToMoveMB}}{\text{throughput}}$$

Notes:
- [1] The maximum number of partition movements given Cruise Control partition movement capacity.

- [2] The network bandwidth given Cruise Control bandwidth throttle.

- [3] The throughput given the max allowed number of partition movements and network bandwidth.

**Estimation for intra-broker rebalance:**

It is challenging to provide an accurate estimate for intra-broker rebalances without an estimate for disk read/write throughput and getting disk throughput is non-trivial for Strimzi.
Since we cannot accurately estimate `estimatedTimeToCompletionInMinutes` without knowing the disk throughput, we set `estimatedTimeToCompletionInMinutes` to `N/A`.

> $$\text{maxPartitionMovements}_{[1]} = \min\left(\text{numberOfBrokers} \times \text{num.concurrent.intra.broker.partition.movements.per.broker}),\text{max.num.cluster.movements}\right)$$

Reviewer 7: Similar here to previous comments, you're missing a bracket in the formula, but I'm also not clear what `maxPartitionMovements` is really representing

    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 95.4347095948149
    onDemandBalancednessScoreBefore: 89.4347095948149
    provisionRecommendation: ""
    provisionStatus: RIGHT_SIZED
    recentWindows: 1
  progress:
    rebalanceProgressConfigMap: my-rebalance-progress
```
[1] Error message from failed Cruise Control REST API call.

### Accessing progress fields using Kubernetes CLI

The progress information will be stored in a `ConfigMap` with the same name as the `KafkaRebalance` resource.
Using the name of the `ConfigMap` we can view its data from the command line using the Kubernetes CLI.

Example accessing `estimatedTimeToCompletionInMinutes` field.
```shell
kubectl get configmaps <my_rebalance_configmap_name> -o json | jq '.["data"]["estimatedTimeToCompletionInMinutes"]'
```

Example accessing `completedDataMovementPercentage` field.
```shell
kubectl get configmaps <my_rebalance_configmap_name> -o json | jq '.["data"]["completedDataMovementPercentage"]'
```

Example accessing `executorState` field.
```shell
kubectl get configmaps <my_rebalance_configmap_name> -o json | jq '.["data"]["executorState"] | fromjson | .'
```

### Affected/not affected projects

> This change will affect the Strimzi Cluster Operator module and mostly the `KafkaRebalanceAssemblyOperator` class.

Reviewer 2: ```suggestion
This change will affect the Strimzi Operators project and mostly the `KafkaRebalanceAssemblyOperator` class in the `cluster-operator` module.
```

[3] The “non-verbose” JSON payload from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint.

[4] The broker load from the optimization proposal as a JSON string that already maintained in the `ConfigMap`.

### Field values per `KafkaRebalance` State

We will provide the progress information for the `KafkaRebalance` states where it is relevant:

- `ProposalReady`:
  - `estimatedTimeToCompletion`: The theoretical minimum estimated time it would take the rebalance to complete if executed.
  - `completedDataMovementPercentage`: 0
  - `executorState`: Empty JSON object.
- `Rebalancing`
  - `estimatedTimeToCompletion`: The estimated time it will take for the ongoing rebalance to complete.
  - `completedDataMovementPercentage`: The percentage of data movement of the ongoing partition rebalance that is complete.
  - `executorState`: JSON object from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint of Cruise Control REST API.
- `Stopped`

>   - `estimatedTimeToCompletion`: "N/A" [1]

Author: I am open too it but worried users might misinterpret `-1` as meaning that the estimated time to completion is in no time, "-1 minutes"! Although having "N/A" forces the type to be a String I feel it would be worth it. That being said, it doesn't matter to me which direction we go with this.

 I would interested in what @ppatierno, he has got a really good eye for this sort of thing.

>   - `estimatedTimeToCompletion`: "N/A" [1]

Reviewer 3: Yeah thinking more and having "N/A" to force the field to be a String is not that great. I was wondering why we cannot use something like we have for the other fields " ... from previous update"? I mean, if it's stopped or it's notready, as we give the completed data movement percentage from the previous update, cannot we use the previous update for the time to completion as well? I the rebalancing stopped, no partitions are moved so it could makes sense as it makes sense for completed data movement. Of course, if the user adds more topics/partitions, the value won't have more sense anymore as well as the one for the completed data movement.

>   - `estimatedTimeToCompletion`: "N/A" [1]

Author: >  I was wondering why we cannot use something like we have for the other fields " ... from previous update"? I mean, if it's stopped or it's notready, as we give the completed data movement percentage from the previous update, cannot we use the previous update for the time to completion as well?

For using the previous values for the `Stopped` and `NotReady` states I worry it may influence users to think the rebalance is still execution or can resumed. However, even if we were to do this we will still need some default error value for instances where the previous value is not available. If we don't feel `N/A` is suitable, I am ok with using `-1`. It is strange enough of a value to raise an alarm IMO and seems to be a common value to express an undefined/error state.

>   - `estimatedTimeToCompletion`: "N/A" [1]

Reviewer 3: > For using the previous values for the Stopped and NotReady states I worry it may influence users to think the rebalance is still execution or can resumed.

why don't you have the same concern on the `completedDataMovementPercentage` field where you are setting the previous update?

>   - `estimatedTimeToCompletion`: "N/A" [1]

Author: > why don't you have the same concern on the `completedDataMovementPercentage` field where you are setting the previous update?

Because `completedDataMovementPercentage` name focuses on what has already happened in the past instead of what may happen in the future. If the field were named something like or `dataToMovePercentage` I would have the same concern as the `estimatedTimeToCompletionInMinutes`.  

That being said, I do see the merits of using the value of the previous update, for one, it would be consistent with how we treat `completedDataMovementPercentage`. Although, I am still slightly leaning in the direction of setting `estimatedTimeToCompletionInMinutes` to an error value for these states,  I am fine with either direction. 

Given my comment here, what do you think? Are you leaning more an either direction?

>   - `estimatedTimeToCompletion`: "N/A" [1]

Reviewer 3: Yes, I am fine with both direction. Let's see what the other thinks and let's make a decision :-)

>   - `estimatedTimeToCompletion`: "N/A" [1]

Reviewer 2: I think you have 3 options -> you can simply leave the field out, or?

>   - `estimatedTimeToCompletion`: "N/A" [1]

Reviewer 3: Thinking more about it, for all reasons mentioned (i.e. field type, client parsing, ...) I think that "N/A" is a no go.
So I would be between -1 and not showing the field at all.
On them, the -1 could be confusing as well, you need to explain and document what exactly it means. So not having it looks like the best option and not confusing people (there is nothing to be confused about).

>   - `estimatedTimeToCompletion`: "N/A" [1]

Author: >   So I would be between -1 and not showing the field at all.
On them, the -1 could be confusing as well, you need to explain and document what exactly it means. So not having it looks like the best option and not confusing people (there is nothing to be confused about).

I have updated the proposal to show what this looks like,  removing the `estimatedTimeToCompletionInMinutes` field from the `progress` section of the `KafkaRebalance` resource for the following states - `ProposalReady`, `Stopped`, `NotReady`. I agree that having this field omitted for these states is better UX ( Why display `estimatedTimeToCompletionInMinutes` in states where the rebalance cannot complete from? ) Every displayed field should serve a purpose whether it be to track the progression of a rebalance or assist in debugging a stopped or failed rebalance. 

Following this reasoning I omitted the `executorState` field from the states where it is undefined as well - `ProposalReady`, `Ready` states. 

This should address the issues raised above, let me know what you think!

  - `completedDataMovementPercentage`: The value of `completedDataMovementPercentage` from previous update.
  - `executorState`: JSON object from previous update.
- `NotReady`
  - `estimatedTimeToCompletion`: "N/A" [1]
  - `completedDataMovementPercentage`: The value of `completedDataMovementPercentage` from previous update.
  - `executorState`: JSON object from previous update.
- `Ready`:
  - `estimatedTimeToCompletion`: 0

>   - `completedDataMovementPercentage`: 100

Author: That is the plan, once the state lands on `Ready`, this value will be hardcoded to `100`.

### Field: `estimatedTimeToCompletionInMinutes`

The estimated time it will take in minutes for a rebalance to complete.

The formulas used to calculate the field value differ for each applicable `KafkaRebalance` state:

#### State: `ProposalReady`

**Estimation for inter-broker rebalance:**

We can only provide an accurate estimate for inter-broker rebalances with an accurate estimate of the network capacity of the brokers.
Therefore, when users do not explicitly set the network capacity settings, `spec.cruiseControl.brokerCapacity.inboundNetwork` and `spec.cruiseControl.brokerCapacity.inboundNetwork` of the `Kafka` resource, `estimatedTimeToCompletionInMinutes` will be "not available" and set `N/A`:

When network capacity values are provided by the user the value of `estimatedTimeToCompletionInMinutes` will be a theoretical minimum derived from these capacity and throttle configurations.
This means that the cluster rebalance would take at least the estimated amount of time to complete.

$$\text{maxConcurrentPartitionMovements}_{[1]} = \min((\text{numberOfBrokers} \times \text{num.concurrent.partition.movements.per.broker}), \text{max.num.cluster.partition.movements})$$

> $$\text{bandwidth}_{[2]} = \min(\frac{\text{networkCapacity}}{\text{maxConcurrentPartitionMovements}}, \text{replication.throttle})$$

Author: > I'm a bit confused, isn't networkCapacity defined per broker in the capacity config? 

If the network capacity is not identical across the brokers we can take the average of the defined network capacity value as our cluster's network capacity as you mentioned below in a later comment.

> Your maxConcurrentPartitionMovements is across the whole cluster isn't it? 

Yes.

> It is also not clear from the CC docs if the replication.throttle is for whole cluster or per broker,

From what I understand from the CC code, it is per broker.

> Even if maxConcurrentPartitionMovements was per broker, what would dividing networkCapacity by it give you? A per partition movement bandwith which is then compared to a per broker bandwidth limit? I don't think those are compatible?

> As networkCapacity is already per broker you just need the minimum of that and the per broker replication.throttle (if set) to get the per-broker bandwidth.

Argh yes you are right, the partition movement bandwidth calculations are unnecessary.

$$\text{throughput}_{[3]} = \text{maxConcurrentPartitionMovements} \times \text{bandwidth}$$

> $$\text{estimatedTimeToCompletionInMinutes} = \frac{\text{dataToMoveMB}}{\text{throughput}}$$

Author: > Basically, total data to move across the whole cluster divided by the per broker bandwidth times the number of brokers. That's it.

This sounds reasonable.

> Also, if I remember correctly, replication throttle can be set by default on the Kafka side as well, so you would need to check if that config is set to be sure of things and doing that from inside the KRAO might be too much effort.

Yes this will also be something to be cognizant of. To my understanding Cruise Control only applies the throttle to the brokers that don't already have a throttle setting configured.

- [3] The throughput given the max allowed number of concurrent partition movements and network bandwidth.

**Estimation for intra-broker rebalance:**

It is challenging to provide an accurate estimate for intra-broker rebalances without an estimate for disk read/write throughput and getting disk throughput is non-trivial for Strimzi.
Since we cannot accurately estimate `estimatedTimeToCompletionInMinutes` without knowing the disk throughput, we set `estimatedTimeToCompletionInMinutes` to `N/A`, an abbreviation meaning "Not Available" or "Not Applicable".

$$\text{maxPartitionMovements}_{[1]} = \min((\text{numberOfBrokers} \times \text{num.concurrent.intra.broker.partition.movements.per.broker}),\text{max.num.cluster.movements})$$

$$\text{estimatedDiskThroughput} = \text{???}_{[2]}$$

$$\text{throughput}_{[3]} = \text{maxPartitionMovements} \times  \text{estimatedDiskThroughput}
$$

$$\text{estimatedTimeToCompletionInMinutes} = \frac{\text{intraBrokerDataToMoveMB}}{\text{throughput}}$$

Since `estimatedDiskThroughput` is unavailable, the formula for `throughput` cannot be resolved, making `estimatedTimeToCompletionInMinutes` undefined or not applicable, `N/A`:

> $$\text{estimatedTimeToCompletionInMinutes} = \text{"N/A"}$$

Author: > Hold on, a rebalance can consist of a combination of inter and intra broker rebalances, what happens then?

From my understanding, a Cruise Control rebalance consists of a single type of partition movement - inter-broker or intra-broker, not a combination of both types.

> $$\text{estimatedTimeToCompletionInMinutes} = \text{"N/A"}$$

Reviewer 3: Yeah I think that's the case AFAIU as well.

Notes
- [1] The maximum number of concurrent partition movements given Cruise Control partition movement capacity.
- [2] We don't have a method of determining disk throughput.
- [3] The throughput given the max allowed number of concurrent partition movements and disk throughput.
 
#### State: `Rebalancing`

The estimated time it will take for the ongoing rebalance to complete based on the average rate of data transfer.

$$

> \text{rate} = \frac{\text{finishedDataMovement}\_{[1]}}{\text{taskTriggerTime}\_{[2]} - \text{currentTime}}

Author: Argh, thanks for spotting this
